src/client.rs

```diff
@@ -233,13 +233,16 @@ fn main() {
                             match chats {
                                 Some(mut chat_messages) => {
                                     chat_messages.retain(|ref chat_message| {
-                                        client_state.chat_msg_seq_num < chat_message.chat_seq.unwrap() &&
-                                          client_state.name.as_ref().unwrap() != &chat_message.player_name
+                                        client_state.chat_msg_seq_num < chat_message.chat_seq.unwrap()
                                     });
+                                    // This loop does two things: 1) update chat_msg_seq_num, and
+                                    // 2) prints messages from other players
                                     for chat_message in chat_messages {
                                         let chat_seq = chat_message.chat_seq.unwrap();
                                         client_state.chat_msg_seq_num = std::cmp::max(chat_seq, client_state.chat_msg_seq_num);
-                                        println!("{}: {}", chat_message.player_name, chat_message.message);
+                                        if client_state.name.as_ref().unwrap() != &chat_message.player_name {
```

Comment:
Why even send back a player's own messages we're just going to ignore it here? Seems like we could use that space for other items in the Update packet, even if it is a minimal saving.

Reply:
@Manghi the synchronization code would be way more complex if we didn't do that (the previous logic where we did not do this was incorrect). If you're concerned about space, what we could do is have an enum like `enum { MyMessage, OtherMessage{ <large_data_here> } }`, so that our own messages take up very little space. The point here is that we have the array elements line up between client and server.

Reply:
Another option would be to have chat IDs shared between client and server, with the array only containing the IDs. We would add a pair of Request/Response messages to the protocol to allow fetching chats by ID.

As an optimization to avoid extra round trips, the Update messages could contain an extra map of chat ID -> chat contents, with the server only adding stuff it thinks the client doesn't have (that is, all new chats except those sent by the client).

Reply:
I really like your second idea. Can we save that as an option if we find this isn't working out? As in, let's move forward with this ("point here is that we have the array elements line up between client and server." convinced me).